be/src/exec/pipeline/adaptive/collect_stats_context.cpp

```diff
@@ -0,0 +1,295 @@
+    const auto& buffer_chunk_queue = _ctx->_buffer_chunk_queue(driver_seq);
+    if (!buffer_chunk_queue.empty()) {
+        return true;
+    }
+
+    size_t num_chunks = _in_chunk_queue_per_driver_seq[driver_seq].size_approx();
+    auto& unpluging = _unpluging_per_driver_seq[driver_seq];
+    if (unpluging) {
+        if (num_chunks > 0) {
+            return true;
+        }
+        unpluging = false;
+        return false;
+    } else if (num_chunks >= UNPLUG_THRESHOLD_PER_DRIVER_SEQ) {
+        unpluging = true;
+        return true;
+    }
+
+    if (_ctx->_is_finishing_per_driver_seq[driver_seq]) {
+        return num_chunks > 0;
+    }
+    return false;
+}
+
+StatusOr<ChunkPtr> PassthroughState::pull_chunk(int32_t driver_seq) {
+    auto& buffer_chunk_queue = _ctx->_buffer_chunk_queue(driver_seq);
+    if (!buffer_chunk_queue.empty()) {
+        auto chunk = std::move(buffer_chunk_queue.front());
+        buffer_chunk_queue.pop();
+        return chunk;
+    }
+
+    auto& passthrough_chunk_queue = _in_chunk_queue_per_driver_seq[driver_seq];
+    ChunkPtr chunk = nullptr;
+    passthrough_chunk_queue.try_dequeue(chunk);
+    return chunk;
+}
+
+Status PassthroughState::set_finishing(int32_t driver_seq) {
+    return Status::OK();
+}
+
+bool PassthroughState::is_source_finished(int32_t driver_seq) const {
+    if (!_ctx->_is_finishing_per_driver_seq[driver_seq]) {
+        return false;
+    }
+
+    const auto& buffer_chunk_queue = _ctx->_buffer_chunk_queue(driver_seq);
+    const auto& passthrough_chunk_queue = _in_chunk_queue_per_driver_seq[driver_seq];
+    return buffer_chunk_queue.empty() && passthrough_chunk_queue.size_approx() <= 0;
+}
+bool PassthroughState::is_sink_finished(int32_t driver_seq) const {
+    return _ctx->_is_finished_per_driver_seq[driver_seq];
+}
+
+/// RoundRobinState.
+RoundRobinState::RoundRobinState(CollectStatsContext* const ctx) : CollectStatsState(ctx) {
+    _info_per_driver_seq.reserve(ctx->_sink_dop);
+    for (int i = 0; i < ctx->_sink_dop; i++) {
+        _info_per_driver_seq.emplace_back(i, _ctx->_runtime_state->chunk_size());
+    }
+}
+
+std::string RoundRobinState::name() const {
+    return "RoundRobin";
+}
+
+bool RoundRobinState::need_input(int32_t driver_seq) const {
+    return false;
+}
+
+Status RoundRobinState::push_chunk(int32_t driver_seq, ChunkPtr chunk) {
+    return Status::InternalError("Shouldn't call RoundRobinState::push_chunk");
+}
+
+bool RoundRobinState::has_output(int32_t driver_seq) const {
+    if (driver_seq >= _ctx->_source_dop) {
+        return false;
+    }
+
+    const auto& [buffer_idx, accumulator] = _info_per_driver_seq[driver_seq];
+    return buffer_idx < _ctx->_sink_dop || !accumulator.empty();
+}
+
+StatusOr<ChunkPtr> RoundRobinState::pull_chunk(int32_t driver_seq) {
+    auto& [buffer_idx, accumulator] = _info_per_driver_seq[driver_seq];
+    if (!accumulator.empty()) {
+        return accumulator.pull();
+    }
+
+    while (buffer_idx < _ctx->_sink_dop) {
```

Comment:
sink_dop /source_dop is quite confusing,  would you like to use upstream_dop, downstream_dop and add some comment to explain that upstream_dop is times of downstream_dop

Reply:
Fine.